packages/liveblocks-react-lexical/src/comments/floating-threads.tsx

```diff
@@ -0,0 +1,321 @@
+        <ThreadWrapper
+          key={thread.id}
+          thread={thread}
+          Thread={Thread}
+          onEscapeKeydown={handleEscapeKeydown}
+          className="lb-lexical-floating-threads-thread"
+        />
+      ))}
+    </FloatingThreadPortal>
+  );
+}
+
+interface FloatingThreadPortalProps
+  extends Omit<HTMLAttributes<HTMLDivElement>, "children"> {
+  range: Range;
+  container: HTMLElement;
+  children: ReactNode;
+}
+
+export const FLOATING_THREAD_COLLISION_PADDING = 10;
+
+function FloatingThreadPortal(props: FloatingThreadPortalProps) {
+  const { container, range, children, className, style, ...divProps } = props;
+
+  const {
+    refs: { setReference, setFloating },
+    strategy,
+    x,
+    y,
+  } = useFloating({
+    strategy: "absolute",
+    placement: "bottom",
+    middleware: [
+      flip({ padding: FLOATING_THREAD_COLLISION_PADDING, crossAxis: false }),
+      offset(10),
+      hide({ padding: FLOATING_THREAD_COLLISION_PADDING }),
+      shift({
+        padding: FLOATING_THREAD_COLLISION_PADDING,
+        limiter: limitShift(),
+      }),
+      size({
+        padding: FLOATING_THREAD_COLLISION_PADDING,
+        apply({ availableWidth, availableHeight, elements }) {
+          elements.floating.style.setProperty(
+            "--lb-lexical-floating-threads-available-width",
+            `${availableWidth}px`
+          );
+          elements.floating.style.setProperty(
+            "--lb-lexical-floating-threads-available-height",
+            `${availableHeight}px`
+          );
+        },
+      }),
+    ],
+    whileElementsMounted: (...args) => {
+      return autoUpdate(...args, {
+        animationFrame: true,
+      });
+    },
+  });
+
+  useLayoutEffect(() => {
+    setReference({
+      getBoundingClientRect: () => range.getBoundingClientRect(),
+    });
+  }, [setReference, range]);
+
+  return createPortal(
+    <div
+      ref={setFloating}
+      {...divProps}
+      style={{
+        ...style,
+        position: strategy,
+        top: 0,
+        left: 0,
+        transform: `translate3d(${Math.round(x)}px, ${Math.round(y)}px, 0)`,
+        minWidth: "max-content",
+      }}
+      className={classNames(
+        "lb-root lb-portal lb-elevation lb-lexical-floating-threads",
+        className
+      )}
+    >
+      {children}
+    </div>,
+    container
+  );
+}
+
+interface ThreadWrapperProps extends ThreadProps {
+  thread: ThreadData;
+  Thread: ComponentType<ThreadProps>;
+  onEscapeKeydown: () => void;
+}
+
+function ThreadWrapper({
+  thread,
+  Thread,
+  style,
+  onEscapeKeydown,
+  onKeyDown,
+  ...threadProps
+}: ThreadWrapperProps) {
+  const handleKeyDown = useCallback(
+    (event: KeyboardEvent<HTMLDivElement>) => {
+      onKeyDown?.(event);
+
+      // TODO: Add ability to preventDefault on keydown to override the default behavior, e.g. to show an alert dialog
+      if (event.key === "Escape") {
+        onEscapeKeydown();
+      }
```

Comment:
We want users to be able to override this behaviour so that they are able to for e.g. display an alert dialog when they press <kbd>Esc</kbd> key.

Reply:
What happens in `FloatingComposer` at the moment?

Reply:
Same behavior unfortunately. The `onKeyDown` event handler on `Composer` is passed to the `form` (and not the Slate.js contenteditable `div`), so the event we get here is a propagated event from the contenteditable div, which is already default prevented. See:

https://github.com/liveblocks/liveblocks/blob/main/packages/liveblocks-react-ui/src/primitives/Composer/index.tsx#L750-L754

Happy to discuss further about this.